### ISSUE TITLE
Test hang on Python connection tests on OS X

### ISSUE THREAD

**Issue by gchpaco** [rethinkdb/rethinkdb#3954]
This came up during work on #2622.

I am experiencing unusual hangs in the testing framework on `next` around connection.py.  Specifically:
```
(rethinkdb-2.7.9)bishop :: rethinkdb/test/rql_test ‹next› » ./test-runner -v connections/connection.py2.7
Warning: did not find any tests for javascript
Warning: did not find any tests for ruby 2.0
Using rethinkdb binary /Users/graham/wd/rethinkdb/build/debug_clang/rethinkdb
	python 2.7 interpreter: /Users/graham/.pyenv/versions/rethinkdb-2.7.9/bin/python2.7, driver: /Users/graham/wd/rethinkdb/drivers/python
== Starting: connections/connection.py2.7 (T+ 1.2 sec)
Running py connection tests
test_auth_key (__main__.TestNoConnection) ... skipped 'Not testing default port'
test_connect (__main__.TestNoConnection) ... skipped 'Not testing default port'
test_connect_host (__main__.TestNoConnection) ... ok
test_connect_port (__main__.TestNoConnection) ... ok
test_connnect_timeout (__main__.TestNoConnection)
Test that we get a ReQL error if we connect to a non-responsive port ... ok
test_empty_run (__main__.TestNoConnection) ... ok
test_connect_correct_auth (__main__.TestAuthConnection) ... ok
test_connect_long_auth (__main__.TestAuthConnection) ... ok
test_connect_no_auth (__main__.TestAuthConnection) ... ok
test_connect_wrong_auth (__main__.TestAuthConnection) ... ok
test_close_does_not_wait_if_requested (__main__.TestConnection) ... ok
test_close_waits_by_default (__main__.TestConnection) ... ok
test_connect_close_expr (__main__.TestConnection) ... ok
test_connect_close_reconnect (__main__.TestConnection) ... ok
test_db (__main__.TestConnection) ... ok
test_noreply_wait_waits (__main__.TestConnection) ... ok
test_port_conversion (__main__.TestConnection) ... ok
test_reconnect_does_not_wait_if_requested (__main__.TestConnection) ... ok
test_reconnect_waits_by_default (__main__.TestConnection) ... ok
test_repl (__main__.TestConnection) ... ok
test_use_outdated (__main__.TestConnection) ... ok
runTest (__main__.TestPrinting) ... ok
runTest (__main__.TestBatching) ... ok
runTest (__main__.TestGetIntersectingBatching) ... Random seed: 13333679129784110953 >>> Timed out connections/connection.py2.7 after 300 seconds after 300.1 sec (T+ 301.4 sec) <<<

== Failed 1 test (of 1) in 301.63 seconds!
	connections/connection.py2.7
```
Here I'm using 2.7.9, but I've seen it in 2.6 as well.

**Comment by Tryneus** [comment]
Just looked into this out of curiosity, and it looks like the server is waiting while part-way through writing a large response to the client (~400KB).  There may be a bug in Tornado's `BaseIOStream.read_bytes`, I doubt this is a problem with the server.  Might try different methods of reading in `TornadoConnection._reader`.

**Comment by gchpaco** [comment]
It happens even in the original connection.py, tho.  I doubt it's a problem with Tornado.

**Comment by Tryneus** [comment]
Ah, then, the problem is probably the server.  Looking into it now.

**Comment by Tryneus** [comment]
It's looking like `linux_event_watcher_t` isn't waking up on the write available event.  'rdhup' isn't implemented for OSX, so it could be that 'out' is broken as well.

**Comment by danielmewes** [comment]
@gchpaco was also looking into this on the client side. But that's interesting. Maybe I broke it with switching OS X to kqueue.

**Comment by gchpaco** [comment]
At a suggestion from @danielmewes I tested using the older `poll_event_queue_t`, which works.  So it would appear that the problem is in `kqueue_event_queue_t`.  Digging.

**Comment by gchpaco** [comment]
The hangup appears to be on this query:
```py
            reference = t1.filter(r.row['geo'].intersects(query_circle))\
                          .coerce_to("ARRAY").run(c)
```
at least *most* of the time.  With 3110841139272090394 as a seed it got as far as the cursor iteration.

**Comment by gchpaco** [comment]
Or maybe not, it's nondeterministic to be sure.

**Comment by gchpaco** [comment]
When I instrumented it it got further, but still hangs on a read halfway through chunking.  If I had to guess, we're not using kqueue properly for sending large chunks.

**Comment by gchpaco** [comment]
To be clearer, it hangs in `SocketWrapper.recvall` when the Python code is expecting a very large read, and gets about ¾ of the way through it in two reads and then hangs indefinitely.  The size is a bit odd (code is saying "saw chunk of size 212044" and then "saw chunk of size 49440" which does not add to any TCP buffer size I know of), but it appears to require multiple sends on the server side, and the server simply isn't sending the remaining data.

**Comment by danielmewes** [comment]
Oh man I had made a stupid bug in the kqueue implementation that would stop it from listening for out/write events if someone was already listening for in/read events on the socket.

Putting the patch up soon.

**Comment by danielmewes** [comment]
Up in CR 2747 by @gchpaco

**Comment by danielmewes** [comment]
Fixed in v1.16.x https://github.com/rethinkdb/rethinkdb/commit/c8d32fc5fb2251bf13311107ed951860cf660d0a , v2.0.x https://github.com/rethinkdb/rethinkdb/commit/55987a649575aa058e0829485897cd69ad2d0d2d and next https://github.com/rethinkdb/rethinkdb/commit/734e6f546e73f530bc54182ef964642bdcdc2363
